File: Quicksort.py
"""
    The program expects a python list in a textfile as input.
    Avg. runtime of Quickssort: O(n*log(n))
"""

import ast

file_path = '../data/numbers.txt'

with open(file_path, 'r') as file:
     numbers = ast.literal_eval(file.read())

def quicksort(l, r):
        if (l < r):
            d = divide(l, r)
            quicksort(l, d)
            quicksort(d+1, r)

def divide(l, r):
    p = numbers[int((l + r) / 2)]
    while (1):
        while (numbers[l] < p):
            l = l+1
        while (numbers[r] > p):
            r = r-1
        if (l < r):
            tmp = numbers[l]
            numbers[l] = numbers[r]
            numbers[r] = tmp
        else:
            return r
        
quicksort(0, len(numbers) -1)

File: Benchmark/src/main/java/Quicksort.java
import java.util.Arrays;

public class Quicksort {

  private int l, r;
  private int[] numbers;

  public static void sort(int[] numbers) {
    quicksort(0, numbers.length-1, numbers);
  }

  public static int divide(int l, int r, int[] numbers) {
    int p = numbers[(l + r) / 2];
      while (true) {
        while (numbers[l] < p)
          l++;
        while (numbers[r] > p)
          r--;
        if (l < r) {
          int tmp = numbers[l];
          numbers[l] = numbers[r];
          numbers[r] = tmp;
        } else
          return r;
      }
  }

  public static void quicksort(int l, int r, int[] numbers) {
    if (l < r) {
      int d = divide(l, r, numbers);
      quicksort(l, d, numbers);
      quicksort(d+1, r, numbers);
    }
  }
}
